src/ripple/app/paths/impl/XRPEndpointStep.cpp

```diff
@@ -0,0 +1,283 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <BeastConfig.h>
+#include <ripple/app/paths/Credit.h>
+#include <ripple/app/paths/impl/AmountSpec.h>
+#include <ripple/app/paths/impl/Steps.h>
+#include <ripple/app/paths/impl/StepChecks.h>
+#include <ripple/basics/Log.h>
+#include <ripple/ledger/PaymentSandbox.h>
+#include <ripple/protocol/IOUAmount.h>
+#include <ripple/protocol/Quality.h>
+#include <ripple/protocol/XRPAmount.h>
+
+
+#include <numeric>
+#include <sstream>
+
+namespace ripple {
+
+class XRPEndpointStep : public StepImp<XRPAmount, XRPAmount, XRPEndpointStep>
+{
+  private:
+    AccountID acc_;
+    bool isLast_;
+    beast::Journal j_;
+
+    boost::optional<XRPAmount> cache_;
+
+  public:
+    XRPEndpointStep (
+        AccountID const& acc,
+        bool isLast,
+        beast::Journal j)
+            :acc_(acc)
+            , isLast_(isLast)
+            , j_ (j) {}
+
+    AccountID const& acc () const
+    {
+        return acc_;
+    };
+
+    boost::optional<EitherAmount>
+    cachedIn () const override
+    {
+        if (!cache_)
+            return boost::none;
+        return EitherAmount (*cache_);
+    }
+
+    boost::optional<EitherAmount>
+    cachedOut () const override
+    {
+        if (!cache_)
```

Comment:
Added a `cached()` helper function that both `cachedIn` and `cachedOut` call and added a comment to `cache~` as to why we use one cache value for both in and out.